### PR TITLE
Fixes regarding explicit names

### PR DESCRIPTION
**Pull request by vouillon**
- only write explicit function names
- when merging modules, the name of types, globals and tags in all modules but the first were lost

### CODE REVIEW

File: src/ir/module-splitting.cpp [hunk 1]
```diff
@@ -542,7 +544,7 @@ void ModuleSplitter::setupTablePatching() {
       placeholder->base = std::to_string(index);
       placeholder->name = Names::getValidFunctionName(
         primary, std::string("placeholder_") + placeholder->base.toString());
-      placeholder->hasExplicitName = false;
+      placeholder->hasExplicitName = true;
```
Comment (vouillon) [inline src/ir/module-splitting.cpp:547]: I'm not sure about this change. Alternatively, we can keep this to false and change a few tests.

File: src/wasm/wasm-binary.cpp [hunk 2]
```diff
@@ -864,19 +864,27 @@ void WasmBinaryWriter::writeNames() {
...
   {
-    auto substart =
-      startSubsection(BinaryConsts::CustomSections::Subsection::NameFunction);
-    o << U32LEB(indexes.functionIndexes.size());
-    Index emitted = 0;
-    auto add = [&](Function* curr) {
-      o << U32LEB(emitted);
-      writeEscapedName(curr->name.str);
-      emitted++;
+    std::vector<std::pair<Index, Function*>> functionsWithNames;
```
Comment (vouillon) [inline src/wasm/wasm-binary.cpp:867]: This is the very same code as for globals.

File: test/unit/input/gc_target_feature.wasm
Comment (vouillon) [inline test/unit/input/gc_target_feature.wasm:1]: The function name section is no longer written when there is no function.

File: src/ir/module-splitting.cpp [hunk 3]
```diff
@@ -542,7 +543,7 @@ void ModuleSplitter::setupTablePatching() {
...
       placeholder->type = secondaryFunc->type;
```
Comment (kripken) [inline src/ir/module-splitting.cpp:547]: Can the `module-splitting` changes be in a separate PR? I believe only wasm-split uses those so it is not relevant for wasm-merge.

File: src/ir/module-utils.cpp [hunk 4]
```diff
@@ -48,6 +48,7 @@ Function* copyFunction(Function* func,
                        std::optional<std::vector<Index>> fileIndexMap) {
   auto ret = std::make_unique<Function>();
   ret->name = newName.is() ? newName : func->name;
+  ret->hasExplicitName = func->hasExplicitName && !newName.is();
```
Comment (kripken) [inline src/ir/module-utils.cpp:51]: Why not set it to true when `newName.is()`?

File: src/ir/module-splitting.cpp (hunk 3 above)
Comment (kripken) [reply src/ir/module-splitting.cpp:547]: Or was the idea to handle all `hasExplicitName` across the codebase? If so then that might make sense but let's remove "wasm-merge" from the PR title.

File: src/ir/module-utils.cpp (hunk 4 above)
Comment (vouillon) [reply src/ir/module-utils.cpp:51]: `newName` is a generated name. For instance, some names like `outlined-A$func-name` are generated during inlining. But maybe we can consider it is explicit enough?

Comment (kripken) [reply src/ir/module-utils.cpp:51]: Ah, I see what you mean, good point. I guess it depends on the user expectation in this code. I think that we do want to keep such names, as we are copying a function, so we should copy the property of having an explicit name.

File: src/ir/module-utils.cpp [hunk 5]
```diff
@@ -48,6 +48,7 @@ Function* copyFunction(Function* func,
                        std::optional<std::vector<Index>> fileIndexMap) {
   auto ret = std::make_unique<Function>();
   ret->name = newName.is() ? newName : func->name;
+  ret->hasExplicitName = newName.is() || func->hasExplicitName;
```
Comment (kripken) [inline src/ir/module-utils.cpp:51]:
```suggestion
  ret->hasExplicitName = func->hasExplicitName;
```
I think it's better to just match the copied function. Their names may differ (if copied into the same module) but the property of having an explicit name seems like it should be the same, at least by default.

File: src/ir/module-splitting.cpp (hunk 1 above)
Comment (kripken) [reply src/ir/module-splitting.cpp:546]: @tlively any thoughts here?

File: src/ir/module-utils.cpp [hunk 6]
```diff
@@ -222,7 +231,6 @@ void copyModule(const Module& in, Module& out) {
   out.customSections = in.customSections;
   out.debugInfoFileNames = in.debugInfoFileNames;
   out.features = in.features;
-  out.typeNames = in.typeNames;
```
Comment (kripken) [inline src/ir/module-utils.cpp:225]: Why did this move? The description of the function it was moved to is that it copies toplevel module items like Functions and Globals, and not metadata.

File: test/unit/input/gc_target_feature.wasm
Comment (kripken) [inline test/unit/input/gc_target_feature.wasm:1]: Why did this change?

File: src/ir/module-utils.cpp (hunk 6 above)
Comment (vouillon) [reply src/ir/module-utils.cpp:225]: So, you think I should rather put the code that copy type names in `wasm-merge.cc`?

File: src/ir/module-splitting.cpp (hunk 1 above)
Comment (vouillon) [reply src/ir/module-splitting.cpp:546]: The thing is that there are a number of tests with an output like:
```
(import "placeholder" "0" (func $placeholder_0 (param i32) (result i32)))
```
If I keep it to `false`, the internal function name is no longer as explicit.

File: test/unit/input/gc_target_feature.wasm
Comment (vouillon) [reply test/unit/input/gc_target_feature.wasm:1]: The function name section is no longer written when there is no function.

File: src/ir/module-splitting.cpp (hunk 1 above)
Comment (tlively) [reply src/ir/module-splitting.cpp:546]: I would be fine keeping this `false` and updating the tests.

File: test/lit/wasm-split/jspi.wast [hunk 7]
```diff
@@ -18,15 +18,15 @@
 
  ;; PRIMARY:      (import "env" "__load_secondary_module" (func $import$__load_secondary_module (param externref)))
 
- ;; PRIMARY:      (import "placeholder" "0" (func $placeholder_0 (param i32) (result i32)))
+ ;; PRIMARY:      (import "placeholder" "0" (func $fimport$1 (param i32) (result i32)))
```
Comment (kripken) [inline test/lit/wasm-split/jspi.wast:21]: This looks like a loss of a useful name. Perhaps it's better to keep it?

File: src/ir/module-utils.cpp (hunk 6 above)
Comment (kripken) [reply src/ir/module-utils.cpp]: Sorry, I don't see why my question, which was about the deletion of this line, is answered by something about `wasm-merge`? What is the context I am missing?

From the standpoint of this file by itself (not thinking about `wasm-merge` at all, this change seems surprising as I said - it looks like this code moved up to `copyModuleItems` which I don't understand the motivation for.

File: test/lit/wasm-split/jspi.wast (hunk 7 above)
Comment (tlively) [reply test/lit/wasm-split/jspi.wast:21]: It's useful in tests, but it's not useful for production. Are tests enough motivation?

Comment (kripken) [reply test/lit/wasm-split/jspi.wast:21]: I think so. In production no names are shipped anyhow so adding more names in non-production builds has little downside, so long as they are meaningful, which looks like the case here.

File: src/ir/module-utils.cpp (hunk 6 above)
Comment (vouillon) [reply src/ir/module-utils.cpp]: `wasm-merge` currently does not copy type names from the input modules to the merged module. I added some code for that in `copyModuleItems`. But then this line became unnecessary.

Comment (kripken) [reply src/ir/module-utils.cpp]: I see, so wasm-merge uses `copyModuleItems` and not `copyModule`. The difference between those two decreases with this change, and the meaning of the difference becomes less clear to me. Perhaps we don't need two functions here at all, and there could just be a single function with maybe a set of options as to whether to copy exports, the start method, etc. (the things only in `copyModule`). Anyhow, for this PR I think the change you made is good, but please add a TODO on `copyModule` to remind us later, maybe something like `TODO: merge this with copyModuleItems, and add options for copying exports and other things that are currently different between them, if we still need those differences`.

**Review by kripken — APPROVED** [review]

**Comment by sbc100** [comment]
This seems to break some emscripten tests:

```
======================================================================
FAIL: test_metadce_hello_O0 (test_other.other.test_metadce_hello_O0)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/test/common.py", line 704, in resulting_test
    return func(self, *args)
           ^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/test/test_other.py", line 8522, in test_metadce_hello
    self.run_metadce_test('hello_world.c', *args)
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/test/test_other.py", line 8457, in run_metadce_test
    self.assertFileContents(filename, data)
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/test/common.py", line 1387, in assertFileContents
    self.assertTextDataIdentical(expected_content, contents, message,
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/test/common.py", line 1345, in assertTextDataIdentical
    return self.assertIdentical(text1, text2, msg, fromfile, tofile)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/test/common.py", line 1364, in assertIdentical
    self.fail(fail_message)
AssertionError: Unexpected difference:
--- /usr/local/google/home/sbc/dev/wasm/emscripten/test/other/metadce/test_metadce_hello_O0.funcs
+++ /usr/local/google/home/sbc/dev/wasm/emscripten/test/other/metadce/test_metadce_hello_O0.funcs.new
@@ -1,3 +1,5 @@
+$54
+$55
 $__DOUBLE_BITS
 $__ashlti3
 $__emscripten_stdout_close
@@ -26,7 +28,6 @@
 $_emscripten_stack_restore
 $_emscripten_tempret_get
 $_emscripten_tempret_set
-$dynCall_jiji
 $emscripten_stack_get_base
 $emscripten_stack_get_current
 $emscripten_stack_get_end
@@ -41,7 +42,6 @@
 $getint
 $getpid
 $init_pthread_self
-$legalstub$dynCall_jiji
 $main
 $memchr
 $out

For full output run with --verbose.
Run with --rebaseline to automatically update expectations
```

**Comment by sbc100** [comment]
Looks like functions that previously had names now just have numbers?

**Comment by sbc100** [comment]
I guess dymCall-generating pass needs to set `hasExplicitName` on the functions it creates?

**Comment by vouillon** [comment]
> I guess dymCall-generating pass needs to set `hasExplicitName` on the functions it creates?

Yes, indeed, that would fix the issue. But I'm wondering whether emscripten really needs the internal names. Maybe you can just update the tests? On the other hand, the generated names are meaningful, so it also make sense to fix that in binaryen.

**Comment by kripken** [comment]
Another fix might be to set `hasExplicitName` by default in the `makeFunction` constructor, when it receives a name. When a string name is provided, at least, we can use it. Functions not given any name at all in wat or wasm binaries are the main situations where we don't want `hasExplicitName` set.

@sbc100 I can look into fixing this but I am out today.

**Comment by sbc100** [comment]
> > I guess dymCall-generating pass needs to set `hasExplicitName` on the functions it creates?
> 
> Yes, indeed, that would fix the issue. But I'm wondering whether emscripten really needs the internal names. Maybe you can just update the tests? On the other hand, the generated names are meaningful, so it also make sense to fix that in binaryen.

The generated names are indeed meaningful/useful, at least for debugging.

**Comment by sbc100** [comment]
> Another fix might be to set `hasExplicitName` by default in the `makeFunction` constructor, when it receives a name. When a string name is provided, at least, we can use it. Functions not given any name at all in wat or wasm binaries are the main situations where we don't want `hasExplicitName` set.
> 

That sounds like a good fix.

> @sbc100 I can look into fixing this but I am out today.

**Comment by vouillon** [comment]
> Another fix might be to set `hasExplicitName` by default in the `makeFunction` constructor, when it receives a name. When a string name is provided, at least, we can use it. Functions not given any name at all in wat or wasm binaries are the main situations where we don't want `hasExplicitName` set.

`makeFunction` is called with a made-up name in several places, though.  For instance, in the binary parser (for imports) and in the legacy text parser.
